Software - Introduction to Serverless workshop (#20)
PoCInnovation/Workshops: software/20.Serverless/README.md

> # Workshop Serverless ✨

Reviewer: Miss workshop number

> Learn the basics of Serverless, an architecture helping you saving time and money.
> ##### ✔️ Serverless App using [Netlify](https://www.netlify.com/), [React](https://fr.reactjs.org/) TS and [Chakra UI](https://chakra-ui.com/)
> ##### ✔️ Serverless API using [Serverless Framework](https://www.serverless.com/)

Reviewer: ```suggestion
✔️ Learn the basics of Serverless, an architecture helping you saving time and money.
✔️ Serverless App using [Netlify](https://www.netlify.com/), [React](https://fr.reactjs.org/) TS and [Chakra UI](https://chakra-ui.com/)
✔️ Serverless API using [Serverless Framework](https://www.serverless.com/)
```

## Introduction 🔰
- You want to stay focus on your code, forget about setting-up severs ?  
- You want to have a scalable application without doing anything ?  
- You want to pay for what you use ?  
- You don't even know what is serverless ?  

> #### Well, you are in the right place at the right time 🍀

Reviewer: Add an explanation, you shouldn't ask questions but instead, answers it.
Check the GRPC and Kafka workshop for examples

> ## Step 0: Setup

Reviewer: ```suggestion
## Step 0 - Setup
```

Watch [this video](https://www.youtube.com/watch?v=tgFiOzVEL0Q) 👀 and please follow the steps in the [SETUP.md](./SETUP.md)  
**Done ? You are now ready to go ! 🔥**

> ## Step 1: Build your app

Reviewer: ```suggestion
## Step 1 - Build your app
```

(Code inside your Serverless Gifs App repository)  

> Click [here](https://eager-albattani-482adb.netlify.app/) to see an example of what you can do 👀  

Reviewer: ```suggestion
[Click to see an example of what you can do](https://eager-albattani-482adb.netlify.app/) 👀  
```

**For now, you will call the giphy API inside the project** 
#### Usefull links:
- [React State Hook](https://fr.reactjs.org/docs/hooks-state.html)
- [React Effect Hook](https://fr.reactjs.org/docs/hooks-effect.html)
- [Chakra UI](https://chakra-ui.com/)
- [iframe](https://developer.mozilla.org/fr/docs/Web/HTML/Element/iframe)

**If you've never learn react, live coding ! 🎥**

## Step 2: Deploy your app
#### Follow these steps:
> First, push your work on Github

> - Commit and push your work

Reviewer: It's the same thing

> > Then you will deploy your application on Netlify

Reviewer: Useless line

- Go to your [Netlify](https://www.netlify.com/) account, `Team overview` and click the `New site from Git` button
  <details>

>     <summary>See the button</summary>

Reviewer: Nothing is deplayed

>    ![](Netlify-App/public/netlifyOverview.png)

Reviewer: Miss description

  </details>
- Select Github for Continious Deployment
  <details>

>     <summary>See the button</summary>

Reviewer: Nothing is displayed

>    ![](Netlify-App/public/netlifyCD.png)

Reviewer: Miss description

  </details>
- Choose your Serverless Gifs App repository

> - Deploy your application with de default settings   

Reviewer: typo

>   **Attention: be aware to the branch to deploy: select your default repository branch**

Reviewer: ```suggestion
  **Warning: be aware to the branch to deploy: select your default repository branch**
```

  <details>

>     <summary>See the default settings</summary>

Reviewer: Nothing is displayed

  
   ![](Netlify-App/public/netlifyDeploy.png)

  </details>
> Good ! Your application is now deployed ! But if you try to use it, we can see that it doesn't work 🤔   

> > Can you guess what the problem is ?  

Reviewer: Miss description

> > Well, you haven't setup your environment variables !  

Reviewer: ```suggestion
> Well, you didn't setup your environment variables!  
```

> Lets do it ⤵️

> - Go to "Site Settings", then "Build and deploy", scroll to the bottom. You will see an "Environment" section

Reviewer: Instead of "" use `Site Settings`

> - Click the "Edit variables" button

Reviewer: Same, don't use  `"` use backtick

  <details>

>     <summary>See the button</summary>

Reviewer: Nothing is displayed

>    ![](Netlify-App/public/netlifyEnv.png)

Reviewer: Miss description

  </details>

> - Add your REACT_APP_GIPHY_API_KEY environment variable and SAVE

Reviewer: Add backtick

> - Go to "Deploys", click the "Trigger Deploy" button and select "Deploy Site"

Reviewer: Backtick

  <details>
    <summary>See the button</summary>
  

>    ![](Netlify-App/public/netlifyTriggerDeploy.png)

Reviewer: Miss description & nothing is displayed

  </details>

> **Your application is now working on production !! 🔥🔥 You can share it 😎**  

Reviewer: How can I test that everything works?

> ## Step 3: Create your first serverless function

Reviewer: ```suggestion
## Step 3 - Create your first serverless function
```

#### Usefull link:
- [Netlify Serverless Functions](https://docs.netlify.com/functions/overview/)

Once you're done, just push your work on your default branch. Netlify will automatically detect changes and will deploy your application.
**Amazing 🎆, isn't it ?**

## Step 4: Create your first serverless API
#### Follow these steps:
- TODO

## Bonus: 

> - #### Create a frontend project calling your serverless API
> - #### Deploy your API with [scaleway](https://www.scaleway.com/en/docs/scaleway-elements-serverless-getting-started/)
>   Why scaleway ? Because there are 🇫🇷 !
> - #### Restart steps 1/2/3, with another API than Giphy

Reviewer: ```suggestion
- Create a frontend project using your serverless API
- Deploy your API with [scaleway](https://www.scaleway.com/en/docs/scaleway-elements-serverless-getting-started/)
  Why scaleway ? Because they are 🇫🇷 !
- Restart steps 1/2/3, with another API than Giphy.
```

## Author

> - [Adrien Fort](https://github.com/adrienfort)

Reviewer: Miss organization, check other workshops for examples
